Adding contrast-limited adaptive histogram equalization (CLAHE) to DALI image operators

tonyreina opened the pull request:
## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:

This PR adds [Contrast-Limited Adaptive Histogram Equalization (CLAHE)](https://docs.opencv.org/4.x/d5/daf/tutorial_py_histogram_equalization.html) to the DALI image operators.

CLAHE performs local histogram equalization with clipping and bilinear blending of lookup tables (LUTs) between neighboring tiles. This technique enhances local contrast while preventing over-amplification of noise. The implementation maintains exact algorithmic compatibility with OpenCV's `cv::createCLAHE()` while providing significant GPU performance optimizations.

## Additional information:

### Affected modules and functionalities:
- Added `clahe_op.cc` and `clahe_op.cu` for GPU implementation with CUDA kernels
- Added `clahe_cpu.cc` for CPU implementation using OpenCV
- Added comprehensive operator schema with detailed documentation
- Added Jupyter Notebook example

### Key points relevant for the review:
- **Algorithmic Compatibility**: Follows the algorithm used by [OpenCV](https://docs.opencv.org/4.x/d5/daf/tutorial_py_histogram_equalization.html)
- **Performance Optimizations**: Includes automatic optimizations:
  - Kernel fusion (RGB→LAB + histogram computation)
  - Warp-privatized histograms for larger tiles (≥1024 pixels)
  - Vectorized memory access for larger images (≥8192 pixels)
  - Adaptive algorithm selection based on image size and tile configuration
- **Feature Support**: 
  - Supports grayscale (1-channel) and RGB (3-channel) uint8 images in HWC layout
  - Two RGB processing modes: luminance-only (preserves color relationships) and per-channel
  - Configurable tile grid, clip limit, and histogram bins

### Tests:
- [x] **New tests added**
  - [x] **Python tests**: `test_clahe.py` with multiple parameter combinations, device testing (CPU/GPU), and API validation
  - [x] **GTests**: `clahe_test.cc` with CPU vs GPU equivalence testing, different tile sizes, clip limits, and error handling
  - [x] **Example**: `clahe_example.py` demonstrating usage patterns and parameter effects
  - [ ] **Benchmark**: Performance benchmarking could be added in future work

## Checklist

### Documentation
- [x] **Documentation updated**
  - [x] **Docstring**: Comprehensive operator schema documentation with parameter descriptions and usage examples
  - [x] **Doxygen**: C++ code documentation following DALI conventions
  - [x] **RST**: Added CLAHE optimizations section to performance tuning guide
  - [x] **Other**: Performance optimization details documented in `PERFORMANCE_NOTES.md`
### DALI team only

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A
<!---  Introduce new or affected requirement IDs, if applicable --->

**JIRA TASK**: N/A
<!--- DALI-XXXX or NA --->

JanuszL commented:
@tonyreina, thank you for your contribution. We appreciate the time you spent diving into DALI and extending it.

I haven't delved deeply into the code yet, as I focused more on general remarks - mostly regarding testing, examples, and memory management. Please let us know if you need any guidance in applying the suggestions.

JanuszL commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1221 @@
+  LaunchHistPerTile256(y_plane, H, W, tiles_x, tiles_y, tmp_histograms, stream);
+  LaunchClipCdfToLut256(tmp_histograms, H, W, tiles_x, tiles_y, clip_limit_rel, tmp_luts, stream);
+  LaunchApplyLUTBilinearToRGB(src_rgb, y_plane, dst_rgb, H, W, tiles_x, tiles_y, tmp_luts, stream);
+  CUDA_CHECK(cudaGetLastError());
```
Please use `CUDA_CALL` which translates the error into a proper exception with the necessary details (see `include/dali/core/cuda_error.h`).

tonyreina commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1221 @@
+             cudaGetErrorString(__err));                                                  \
+    }                                                                                     \
+  } while (0)
+#endif
```
Sorry. Didn't realize that existed.

mzient commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1190 @@
+
+static inline __host__ __device__ int div_up(int a, int b) {
+  return (a + b - 1) / b;
+}
```
Nitpick: we already have it implemented as `div_ceil` in `dali/core/util.h`.

jantonguirao commented on dali/operators/image/clahe/clahe_cpu.cc:
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
General nitpick: Please use 2025 in copyright headers

jantonguirao commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1190 @@
+    z_xyz = z_xyz / 1.088754f;
+
+    // Convert Y to LAB L* using OpenCV's threshold and constants
+    float fy = (y_xyz > 0.008856f) ? cbrtf(y_xyz) : (7.787f * y_xyz + 16.0f / 116.0f);
```
You used `008856f` here but 
```
  const float delta = 6.0f / 29.0f;
  const float delta_cube = delta * delta * delta;
```
in `xyz_to_lab_f`, which is way more precise. Please use a helper function in both places which uses the accurate constant.

jantonguirao commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1190 @@
+
+    // Tile geometry - use same calculation as histogram kernel for consistency
+    int tile_w = div_up(W, tiles_x);
+    int tile_h = div_up(H, tiles_y);
```
Those could be calculated once outside of the loop.

jantonguirao commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1190 @@
+    int y = y0 + dy;
+
+    int pixel_idx = y * W + x;
+    int rgb_idx = 3 * pixel_idx;
```
```suggestion
    size_t pixel_idx = static_cast<size_t>(y) * W + x;
    size_t rgb_idx = 3 * pixel_idx;
```
To avoid overflow.

jantonguirao commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1190 @@
+    float v_out = v_top * (1.f - fy) + v_bot * fy;
+
+    int outi = static_cast<int>(lrintf(fminf(fmaxf(v_out, 0.f), 255.f)));
+    dst_y[idx] = (uint8_t)outi;
```
Same here. You could benefit from ConvertSat<uint8_t>(...)

jantonguirao commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1190 @@
+  float v_out = v_top * (1.f - fy) + v_bot * fy;
+
+  int outi = static_cast<int>(lrintf(fminf(fmaxf(v_out, 0.f), 255.f)));
+  dst_y[idx] = (uint8_t)outi;
```
ConvertSat<uint8_t>(...) would read better here.

Same apply below (won't repeat each time)

jantonguirao commented on dali/test/python/operator_1/test_clahe.py:
```diff
@@ -0,0 +1,242 @@
+
+    def define_graph(self):
+        # Create synthetic test data
+        data = fn.random.uniform(range=(0, 255), shape=self.input_shape, dtype=types.UINT8)
```
please set the random seed so that we can have reproducible results.

jantonguirao commented on dali/test/python/operator_1/test_clahe.py:
```diff
@@ -0,0 +1,242 @@
+    # Verify output properties
+    assert len(clahe_output) == batch_size
+    for i in range(batch_size):
+        original = np.array(input_data[i])
```
I think you need to add `.as_cpu()` before constructing the np.array. Otherwise it won't work in case of GPU outputs
```suggestion
        original = np.array(input_data[i].as_cpu())
```
Same applies to all conversions to np.array in this file.

jantonguirao commented on dali/test/python/operator_1/test_clahe.py:
```diff
@@ -0,0 +1,242 @@
+        for i in range(batch_size):
+            enhanced = np.array(clahe_output[i])
+            assert enhanced.shape == (64, 64, 1)
+            assert enhanced.dtype == np.uint8
```
The tests only check the shape and type.
How about comparing against some reference implementation such as OpenCV, to check correctness?

jantonguirao commented:
Would be nice to add images to the example, to see the effect of this transformation visually

dali-automaton commented:
**CI MESSAGE**: [36715300]: BUILD FAILED

tonyreina commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1190 @@
+  int area = (x1 - x0) * (y1 - y0);
+  for (int i = threadIdx.x; i < area; i += blockDim.x) {
+    int dy = i / (x1 - x0);
+    int dx = i - dy * (x1 - x0);
```
I think `int dy = i / (x1 - x0)` calculates the row offset within the tile, but `int dy = i % (x1 - x0)` calculates the column offset within the tile. I tried changing to modulo, but it generates a big error in the output image.

JanuszL replied:
I think the suggestion was:
```suggestion
    int dy = i / (x1 - x0);
    int dx = i % (x1 - x0);
```

dali-automaton commented:
**CI MESSAGE**: [36728551]: BUILD FAILED

dali-automaton commented:
**CI MESSAGE**: [36780586]: BUILD FAILED

tonyreina commented:
> Would be nice to add images to the example, to see the effect of this transformation visually

I updated the notebook with a test image to show the CLAHE processing.

tonyreina commented:
I merged the recent changes changes from `main` about an hour ago. On my system I just build by going to `cd DALI/docker; ./build.sh` which grabs the Docker and builds the package inside the Docker. That's been working really well for me.

After the latest merge from `main` I can no longer get all the way through the `build.sh`. It starts complaining about bundling `ffmpeg` in the `pip` wheel:

```
$ORIGIN $ORIGIN $ORIGIN $ORIGIN Fixed rpath of .lib files Generating new record file nvidia_dali_cuda130-1.53.0.dev0.dist-info/RECORD Removing lock file... Finished generating new record file nvidia_dali_cuda130-1.53.0.dev0.dist-info/RECORD Check bundled libs... ERROR: The library: 'libavformat.so.61' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavcodec.so.61' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavfilter.so.10' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavutil.so.59' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libswscale.so.8' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavformat.so.61' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavcodec.so.61' should be bundled in whl or removed from the dynamic link dependency
```

I checked out a fresh version of `main` (i.e. not my fork) and see the same error.

Can anyone confirm?

My guess is this:  https://github.com/NVIDIA/DALI/commit/e99864f51042079ffabe4d9d4a2e99ceb9b429bb#diff-83178761b5c58921837b9674b588e74d0238509411d2a5ba5eb9b003bb33a1b1

JanuszL commented:
> I merged the recent changes changes from `main` about an hour ago. On my system I just build by going to `cd DALI/docker; ./build.sh` which grabs the Docker and builds the package inside the Docker. That's been working really well for me.
> 
> After the latest merge from `main` I can no longer get all the way through the `build.sh`. It starts complaining about bundling `ffmpeg` in the `pip` wheel:
> 
> ```
> $ORIGIN $ORIGIN $ORIGIN $ORIGIN Fixed rpath of .lib files Generating new record file nvidia_dali_cuda130-1.53.0.dev0.dist-info/RECORD Removing lock file... Finished generating new record file nvidia_dali_cuda130-1.53.0.dev0.dist-info/RECORD Check bundled libs... ERROR: The library: 'libavformat.so.61' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavcodec.so.61' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavfilter.so.10' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavutil.so.59' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libswscale.so.8' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavformat.so.61' should be bundled in whl or removed from the dynamic link dependency ERROR: The library: 'libavcodec.so.61' should be bundled in whl or removed from the dynamic link dependency
> ```
> 
> I checked out a fresh version of `main` (i.e. not my fork) and see the same error.
> 
> Can anyone confirm?
> 
> My guess is this: [e99864f#diff-83178761b5c58921837b9674b588e74d0238509411d2a5ba5eb9b003bb33a1b1](https://github.com/NVIDIA/DALI/commit/e99864f51042079ffabe4d9d4a2e99ceb9b429bb#diff-83178761b5c58921837b9674b588e74d0238509411d2a5ba5eb9b003bb33a1b1)

Hi @tonyreina,

Sorry for causing this troubles. We updated the FFmpeg version in the builder image (which is not rebuild automatically to save your time) and adjusted scripts in the main repo. Please run ` REBUILD_BUILDERS=YES ./build.sh` to make builder image with sync with the rest of the repository.

tonyreina commented:
> Hi @tonyreina,
> 
> Sorry for causing this troubles. We updated the FFmpeg version in the builder image (which is not rebuild automatically to save your time) and adjusted scripts in the main repo. Please run ` REBUILD_BUILDERS=YES ./build.sh` to make builder image with sync with the rest of the repository.

Yes. That was the issue! Thanks for helping. It builds correctly now on my computer.

Is it still giving build errors in the tests? Is there any way for me to run those in the `./build.sh` so that I can see the error trace? I'm guessing it is something I did in updating the python test scripts rather than the CLAHE code itself.

```bash
pixi run pytest ~/DALI/dali/test/python/operator_1/test_clahe.py -v
====================================================================================== test session starts =======================================================================================
platform linux -- Python 3.12.11, pytest-8.4.2, pluggy-1.6.0 -- /home/exthymic/dali_tracking_pipeline/.pixi/envs/default/bin/python3.12
cachedir: .pytest_cache
rootdir: /home/exthymic/DALI
configfile: pyproject.toml
plugins: anyio-4.10.0, hydra-core-1.3.2
collected 11 items                                                                                                                                                                               

../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_operator_registration PASSED                                                                                                 [  9%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_grayscale_gpu PASSED                                                                                                         [ 18%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_rgb_gpu PASSED                                                                                                               [ 27%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_ops_api PASSED                                                                                                               [ 36%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_parameter_validation PASSED                                                                                                  [ 45%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_different_tile_configurations PASSED                                                                                         [ 54%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_opencv_comparison_gpu PASSED                                                                                                 [ 63%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_opencv_comparison_cpu PASSED                                                                                                 [ 72%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_gpu_cpu_consistency PASSED                                                                                                   [ 81%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_different_parameters_accuracy PASSED                                                                                         [ 90%]
../DALI/dali/test/python/operator_1/test_clahe.py::test_clahe_medical_image_accuracy PASSED                                                                                                [100%]

======================================================================================= 11 passed in 3.02s =======================================================================================
```

jantonguirao commented:
@tonyreina Thank you for adding the example. It looks good now!
I only see one potential unresolved problem here: https://github.com/NVIDIA/DALI/pull/6069/files#r2435357237
Could you please take a look? Thanks!

jantonguirao commented:
!build

dali-automaton commented:
**CI MESSAGE**: [36909584]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [36909584]: BUILD FAILED

JanuszL commented:
!build

JanuszL commented:
Hi @tonyreina,

> Is nose2 the correct testing package for python?

Yes, it is. You need to install it separately (together with other packages necessary for tests) as the builder image doesn't have these preinstalled.

> Should I mount the Docker /opt/dali to the git repository root or the build directory root?

Mount the git repo root to /opt/dali

dali-automaton commented:
**CI MESSAGE**: [36989671]: BUILD STARTED

JanuszL commented:
@tonyreina what I see in CI is:
```
[2025-10-21 17:39:42] ======================================================================
[2025-10-21 17:39:42] FAIL: test_eager_coverage.test_clahe
[2025-10-21 17:39:42] ----------------------------------------------------------------------
[2025-10-21 17:39:42] Traceback (most recent call last):
[2025-10-21 17:39:42]   File "/usr/local/lib/python3.10/dist-packages/nose/case.py", line 198, in runTest
[2025-10-21 17:39:42]     self.test(*self.arg)
[2025-10-21 17:39:42]   File "/opt/dali/dali/test/python/test_eager_coverage.py", line 661, in test_clahe
[2025-10-21 17:39:42]     check_single_input("clahe", tiles_x=4, tiles_y=4, clip_limit=2.0)
[2025-10-21 17:39:42]   File "/opt/dali/dali/test/python/test_eager_coverage.py", line 227, in check_single_input
[2025-10-21 17:39:42]     compare_eager_with_pipeline(pipe, eager_op, eager_source=eager_source, layout=layout, **kwargs)
[2025-10-21 17:39:42]   File "/opt/dali/dali/test/python/test_eager_coverage.py", line 200, in compare_eager_with_pipeline
[2025-10-21 17:39:42]     check_batch(tensor_out_fn, tensor_out_eager, batch_size)
[2025-10-21 17:39:42]   File "/opt/dali/dali/test/python/test_utils.py", line 361, in check_batch
[2025-10-21 17:39:42]     assert False, error_msg
[2025-10-21 17:39:42] AssertionError: Mean error: [7.289166666666667], Min error: [0], Max error: [82]
[2025-10-21 17:39:42] Total error count: [634], Tensor size: [1200]
[2025-10-21 17:39:42] Index in batch: 0
[2025-10-21 17:39:42] 
[2025-10-21 17:39:42] LHS data source: 
[2025-10-21 17:39:42] RHS data source: 
```
I hope this helps you narrow down the problem.

tonyreina commented on dali/operators/image/clahe/clahe_cpu.cc:
```diff
@@ -0,0 +1,137 @@
+    const auto &in = ws.Input<CPUBackend>(0);
+
+    if (in.type() != DALI_UINT8) {
+      throw std::invalid_argument("ClaheCPU currently supports only uint8 input.");
```
I think you wanted me to use the std library instead of these, but I'm happy to move to this style if necessary

tonyreina commented on dali/test/python/operator_1/test_clahe.py:
```diff
@@ -0,0 +1,594 @@
+                # Bright block with some detail
+                img[i : i + block_size, j : j + block_size] = np.random.randint(
+                    200, 220, (block_size, block_size, 3)
+                )
```
Done

tonyreina commented on dali/test/python/operator_1/test_clahe.py:
```diff
@@ -0,0 +1,594 @@
+    try:
+        # Create memory-based pipeline
+        pipe = MemoryPipeline(image_array, tiles_x, tiles_y, clip_limit, device)
+        pipe.build()
```
Added it.

tonyreina commented on dali/operators/image/clahe/clahe_test.cc:
```diff
@@ -0,0 +1,218 @@
+                         .AddInput("input", StorageDevice::CPU)
+                         .AddOutput("output", StorageDevice::CPU));
+  });
+}
```
added more tests

tonyreina commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1056 @@
+  // https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/color_lab.cpp#L1023
+  float cf = c / 255.0f;
+  return (cf <= GAMMA_THRESHOLD) ? cf / GAMMA_LOW_SCALE :
+                                   powf((cf + GAMMA_XSHIFT) / (1.0f + GAMMA_XSHIFT), GAMMA_POWER);
```
I'd rather keep this because it is the original equation that OpenCV uses.

tonyreina commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1056 @@
+    float val = static_cast<float>(cdf[i]) * lutScale + 0.5f;
+    lut[i] = static_cast<uint8_t>(dali::clamp(val, 0.f, 255.f));
+  }
+}
```
Ok. Fixed.

tonyreina commented on dali/operators/image/clahe/clahe_op.cc:
```diff
@@ -0,0 +1,207 @@
+      auto sample_shape = shape.tensor_shape_span(i);
+      if (sample_shape.size() != 3) {
+        throw std::invalid_argument("ClaheGPU expects HWC input layout.");
+      }
```
updated

tonyreina commented on dali/operators/image/clahe/clahe_op.cc:
```diff
@@ -0,0 +1,207 @@
+      int C = sample_shape[2];
+      if (C != 1 && C != 3) {
+        throw std::invalid_argument("ClaheGPU supports 1 or 3 channels.");
+      }
```
same as previous

tonyreina commented on dali/test/python/operator_1/test_clahe.py:
```diff
@@ -0,0 +1,594 @@
+    def define_graph(self):
+        # Create synthetic test data - CLAHE requires uint8 input
+        data = fn.cast(
+            fn.random.uniform(range=(0, 255), shape=self.input_shape, seed=816),
```
I think this should be ok.

tonyreina commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1056 @@
+  for (int i = 0; i < 2; ++i) {
+    int idx = base_idx + i;
+    if (idx >= N) {
+      return;
```
updated

tonyreina commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1056 @@
+  for (int i = 0; i < 4; ++i) {
+    int idx = base_idx + i;
+    if (idx >= N) {
+      return;
```
updated

tonyreina commented on dali/operators/image/clahe/clahe_cpu.cc:
```diff
@@ -0,0 +1,138 @@
+    auto &shape = in_sample.shape;
+    int H = shape[0];
+    int W = shape[1];
+    int C = (shape.size() >= 3) ? shape[2] : 1;
```
updated

tonyreina commented:
I think the line 87 is a false positive because If rank-2 (grayscale HW format), it defaults to C=1. That should be ok.

tonyreina commented:
Hopefully the new build will work correctly with the tests. crossing fingers!

dali-automaton commented:
**CI MESSAGE**: [37566374]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [37566374]: BUILD FAILED

JanuszL commented:
!build

tonyreina commented:
<img width="1746" height="1113" alt="image" src="https://github.com/user-attachments/assets/8440e242-6f0d-4d35-840a-cb93984b88b9" />

-- | -- | --
docs/examples/image_processing/clahe_example.ipynb | 1/5 | Empty file containing only newline - critical documentation missing
-- | -- | --


<!--EndFragment-->
</body>
</html>

greptile is reporting that the notebook is empty, but I can see the notebook contents. Maybe it is a false positive? Is it not able to parse notebooks?

Why did the last build fail? The previous one passed. I didn't change anything on my end.

mzient commented on dali/test/python/operator_1/test_clahe.py:
```diff
@@ -0,0 +1,586 @@
+            del pipe
+
+
+class ClahePipeline(Pipeline):
```
Wouldn't it be easier to use `@pipeline_def` decorator instead of inherting from `Pipeline`?

mzient commented on dali/test/python/test_dali_variable_batch_size.py:
```diff
@@ -346,7 +357,10 @@ def numba_setup_out_shape(out_shape, in_shape):
+    (
+        fn.fast_resize_crop_mirror,
+        {"crop": [5, 5], "resize_shorter": 10, "devices": ["cpu"]},
+    ),
```
Nitpick: that's not what our formatting tools produce - we have some config for `black`, extending the line length to 100 characters. Please revert the formatting changes and re-run black with the proper config (I think we have black configuration in our repo).

mzient commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1207 @@
+
+  // Compute both paths to avoid warp divergence
+  float linear_path = cf / GAMMA_LOW_SCALE;
+  float gamma_path = powf((cf + GAMMA_XSHIFT) / (1.0f + GAMMA_XSHIFT), GAMMA_POWER);
```
Again, you could use multiplication by reciprocal.

mzient commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1207 @@
+  // Use predication (no branch divergence)
+  float use_cbrt = (t > LTHRESHOLD) ? 1.0f : 0.0f;
+  return use_cbrt * cbrt_path + (1.0f - use_cbrt) * linear_path;
+}
```
Same remarks as above.

mzient commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1207 @@
+
+  // Use predication (no branch divergence)
+  float use_cubic = (u > THRESHOLD_6_29TH) ? 1.0f : 0.0f;
+  return use_cubic * cubic_path + (1.0f - use_cubic) * linear_path;
```
..and again.

tonyreina commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1207 @@
+#include "dali/core/math_util.h"
+#include "dali/core/util.h"
+
+#define CV_HEX_CONST_F(x) static_cast<float>(__builtin_bit_cast(double, (uint64_t)(x)))
```
I updated to using `std::bit_cast` but the `##_u64` caused a compile error. Would this update be ok?

```
// CUDA-compatible bit_cast: use C++20 std::bit_cast in host code,
// union-based reinterpretation in device code
#ifdef __CUDA_ARCH__
// Device code: use union for bit reinterpretation
__device__ inline double uint64_to_double(uint64_t x) {
  union {
    uint64_t u;
    double d;
  } converter;
  converter.u = x;
  return converter.d;
}
#else
// Host code: use C++20 std::bit_cast
inline double uint64_to_double(uint64_t x) {
  return std::bit_cast<double>(x);
}
#endif

#define CV_HEX_CONST_F(x) static_cast<float>(uint64_to_double(static_cast<uint64_t>(x)))
```

mzient replied:
Nah, if the __builtin_bit_cast worked in device code, let's keep it this way and we'll then work on unification efforts.

mzient commented on dali/operators/image/clahe/clahe_op.cu:
```diff
@@ -0,0 +1,1207 @@
+// Histogram clipping, redistribution, and CDF calculation helper
+// -------------------------------------------------------------------------------------
+
+__device__ void clip_redistribute_cdf(unsigned int *h, int bins, int area, float clip_limit_rel,
```
Optimizing this piece is likely out of scope of this initial PR, so please mark it with some `TODO` that this code needs some attention. Currently it has a lot of sequential computations involving global memory (lut) and it's amenable to parallelization, at least at warp level.

tonyreina commented:
I think I made all of the recommended changes. The notebook and test examples now use the MRI knee image from DALI_extra.

tonyreina commented:
I updated the powf calls with a LUT to improve speed. The other greptile comment I think was around the parallelization of the CDF calculation. Happy to add that if it is really necessary.

dali-automaton commented:
**CI MESSAGE**: [38318833]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [38318833]: BUILD FAILED

dali-automaton commented:
**CI MESSAGE**: [38418710]: BUILD STARTED

JanuszL commented:
!build

tonyreina commented:
> I see following errors: Gtest:
> 
> ```
> [2025-11-14 09:47:15] [ RUN      ] ClaheOpTest.LumaOnlyVsPerChannel
> [2025-11-14 09:47:15] 
> [2025-11-14 09:47:15] !!! Forced use of Executor 2.0 !!!
> [2025-11-14 09:47:15] 
> [2025-11-14 09:47:15] !!! Forced use of Executor 2.0 !!!
> [2025-11-14 09:47:15] 
> [2025-11-14 09:47:15] !!! Forced use of Executor 2.0 !!!
> [2025-11-14 09:47:15] 
> [2025-11-14 09:47:15] !!! Forced use of Executor 2.0 !!!
> [2025-11-14 09:47:15] unknown file: Failure
> [2025-11-14 09:47:15] C++ exception with description "Critical error in pipeline:
> [2025-11-14 09:47:15] Critical error in pipeline:
> [2025-11-14 09:47:15] Error in GPU operator `nvidia.dali.ops.Clahe` encountered:
> [2025-11-14 09:47:15] 
> [2025-11-14 09:47:15] ClaheGPU RGB input requires luma_only=True; per-channel mode (luma_only=False) is not supported on GPU. Use CPU backend for per-channel processing.
> [2025-11-14 09:47:15] Current pipeline object is no longer valid.
> [2025-11-14 09:47:15] Current pipeline object is no longer valid." thrown in the test body.
> [2025-11-14 09:47:15] 
> ```
> 
> test_clahe.py
> 
> ```
> [2025-11-14 11:02:30] ======================================================================
> [2025-11-14 11:02:30] ERROR: test_clahe (nose2.loader.ModuleImportFailure)
> [2025-11-14 11:02:30] ----------------------------------------------------------------------
> [2025-11-14 11:02:30] ImportError: Failed to import test module: test_clahe
> [2025-11-14 11:02:30] Traceback (most recent call last):
> [2025-11-14 11:02:30]   File "/usr/local/lib/python3.10/dist-packages/nose2/plugins/loader/discovery.py", line 195, in _find_tests_in_file
> [2025-11-14 11:02:30]     module = util.module_from_name(module_name)
> [2025-11-14 11:02:30]   File "/usr/local/lib/python3.10/dist-packages/nose2/util.py", line 73, in module_from_name
> [2025-11-14 11:02:30]     __import__(name)
> [2025-11-14 11:02:30]   File "/opt/dali/dali/test/python/operator_1/test_clahe.py", line 17, in <module>
> [2025-11-14 11:02:30]     import pytest
> [2025-11-14 11:02:30] ModuleNotFoundError: No module named 'pytest
> '```
> ```

ok. please try the build again.

thanks!

-Tony

dali-automaton commented:
**CI MESSAGE**: [38651866]: BUILD FAILED

JanuszL commented:
@tonyreina 
It seems that only one issue remains:
```
[2025-11-17 22:00:06] [ RUN      ] ClaheOpTest.GpuRejectsPerChannelMode
[2025-11-17 22:00:06] 
[2025-11-17 22:00:06] !!! Forced use of Executor 2.0 !!!
[2025-11-17 22:00:06] /opt/dali/dali/operators/image/clahe/clahe_test.cc:202: Failure
[2025-11-17 22:00:06] Expected: pipe.Run() throws an exception of type std::runtime_error.
[2025-11-17 22:00:06]   Actual: it throws nothing.```

tonyreina commented:
I think it couldn't be scheduled for the build.

dali-automaton commented:
**CI MESSAGE**: [38766317]: BUILD STARTED

dali-automaton commented:
**CI MESSAGE**: [38766317]: BUILD PASSED

JanuszL commented:
@tonyreina the CI is passing so please let us review the latest changes again. We hope to do that soon (it is over 4k lines of code after all).
